File: src/charting/highlight/CombinedHighlighter.java
package com.github.mikephil.charting.highlight;

import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarLineScatterCandleBubbleData;
import com.github.mikephil.charting.data.ChartData;
import com.github.mikephil.charting.data.DataSet;
import com.github.mikephil.charting.interfaces.dataprovider.BarDataProvider;
import com.github.mikephil.charting.interfaces.dataprovider.CombinedDataProvider;
import com.github.mikephil.charting.interfaces.datasets.IDataSet;

import java.util.List;

/**
 * Created by Philipp Jahoda on 12/09/15.
 */
public class CombinedHighlighter extends ChartHighlighter<CombinedDataProvider> implements IHighlighter
{

    /**
     * bar highlighter for supporting stacked highlighting
     */
    protected BarHighlighter barHighlighter;

    public CombinedHighlighter(CombinedDataProvider chart, BarDataProvider barChart) {
        super(chart);

        // if there is BarData, create a BarHighlighter
        barHighlighter = barChart.getBarData() == null ? null : new BarHighlighter(barChart);
    }

    
    protected List<Highlight> getHighlightsAtXValue(let xVal, let x, let y) {

        this.mHighlightBuffer.clear();

        List<BarLineScatterCandleBubbleData> dataObjects = this.mChart.getCombinedData().getAllData();

        for (let i = 0; i < dataObjects.length; i++) {

            ChartData dataObject = dataObjects.get(i);

            // in case of BarData, let the BarHighlighter take over
            if (barHighlighter != null && dataObject instanceof BarData) {
                Highlight high = barHighlighter.getHighlight(x, y);

                if (high != null) {
                    high.setDataIndex(i);
                    this.mHighlightBuffer.add(high);
                }
            } else {

                for (let j = 0, dataSetCount = dataObject.getDataSetCount(); j < dataSetCount; j++) {

                    IDataSet dataSet = dataObjects.get(i).getDataSetByIndex(j);

                    // don't include datasets that cannot be highlighted
                    if (!dataSet.isHighlightEnabled())
                        continue;

                    List<Highlight> highs = buildHighlights(dataSet, j, xVal, DataSet.Rounding.CLOSEST);
                    for (Highlight high : highs)
                    {
                        high.setDataIndex(i);
                        this.mHighlightBuffer.add(high);
                    }
                }
            }
        }

        return this.mHighlightBuffer;
    }

//    protected Highlight getClosest(let x, let y, Highlight... highs) {
//
//        Highlight closest = null;
//        let minDistance = Infinity;
//
//        for (Highlight high : highs) {
//
//            if (high == null)
//                continue;
//
//            let tempDistance = getDistance(x, y, high.getXPx(), high.getYPx());
//
//            if (tempDistance < minDistance) {
//                minDistance = tempDistance;
//                closest = high;
//            }
//        }
//
//        return closest;
//    }
}

File: src/charting/highlight/CombinedHighlighter.ts
import { ChartHighlighter } from './ChartHighlighter';
import { BarHighlighter } from './BarHighlighter';
import { IHighlighter } from './IHighlighter';
import { CombinedDataProvider } from '../interfaces/dataprovider/CombinedDataProvider';
import { BarDataProvider } from '../interfaces/dataprovider/BarDataProvider';
import { BarData } from '../data/BarData';
import { Rounding } from '../data/DataSet';
/**
 * Created by Philipp Jahoda on 12/09/15.
 */
export class CombinedHighlighter extends ChartHighlighter<CombinedDataProvider> implements IHighlighter {
    /**
     * bar highlighter for supporting stacked highlighting
     */
    protected barHighlighter: BarHighlighter;

    constructor(chart: CombinedDataProvider, barChart: BarDataProvider) {
        super(chart);

        // if there is BarData, create a BarHighlighter
        this.barHighlighter = barChart.barData == null ? null : new BarHighlighter(barChart);
    }

    public getHighlightsAtXValue(xVal, x?, y?) {
        this.mHighlightBuffer = [];

        const dataObjects = this.mChart.combinedData.getAllData();

        for (let i = 0; i < dataObjects.length; i++) {
            const dataObject = dataObjects[i];

            // in case of BarData, let the BarHighlighter take over
            if (this.barHighlighter != null && dataObject instanceof BarData && x !== undefined) {
                const high = this.barHighlighter.getHighlight(x, y);

                if (high != null) {
                    high.dataIndex = i;
                    this.mHighlightBuffer.push(high);
                }
            } else {
                for (let j = 0, dataSetCount = dataObject.dataSetCount; j < dataSetCount; j++) {
                    const dataSet = dataObjects[i].getDataSetByIndex(j);

                    // don't include datasets that cannot be highlighted
                    if (!dataSet.highlightEnabled) continue;

                    const highs = this.buildHighlights(dataSet, j, xVal, Rounding.CLOSEST);
                    for (const high of highs) {
                        high.dataIndex = i;
                        this.mHighlightBuffer.push(high);
                    }
                }
            }
        }

        return this.mHighlightBuffer;
    }

    //    protected Highlight getClosest(let x, let y, Highlight... highs) {
    //
    //        Highlight closest = null;
    //        let minDistance = Infinity;
    //
    //        for (Highlight high : highs) {
    //
    //            if (high == null)
    //                continue;
    //
    //            let tempDistance = getDistance(x, y, high.getXPx(), high.getYPx());
    //
    //            if (tempDistance < minDistance) {
    //                minDistance = tempDistance;
    //                closest = high;
    //            }
    //        }
    //
    //        return closest;
    //    }
}
